### PR TITLE
gstreamer: use monorepo build

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Upstream have merged all their projects into a monorepo. [1] The
supported way of building the various projects is through the monorepo,
so we should do that.

See Homebrew/discussions#3740 for additional context.

Resolves Homebrew/discussions#3740.

[1] https://gstreamer.freedesktop.org/documentation/frequently-asked-questions/mono-repository.html?gi-language=c

### CODE REVIEW

**Comment by chenrui333** [comment]
💯

**Comment by carlocab** [comment]
```
OSError: [Errno 86] Bad CPU type in executable: '/private/tmp/gstreamer-20230318-31006-a2njdq/gstreamer-1.22.1/subprojects/macos-bison-binary/bison-3.7.6-macos-x86_64/bin/bison'
```
Why are you downloading your own `bison`? How annoying.

**Comment by carlocab** [comment]
```
  error[E0069]: `return;` in a function whose return type is not `()`
     --> video/gtk4/src/sink/imp.rs:856:21
      |
  856 |                     return;
      |                     ^^^^^^ return type is not `()`
  
  For more information about this error, try `rustc --explain E0069`.
```
Not sure why this didn't fail when building `gst-plugins-rs` previously, though.

File: Formula/gstreamer.rb [hunk 1]
```diff
@@ -1,10 +1,37 @@
...
+    url "https://gitlab.freedesktop.org/gstreamer/gstreamer/-/archive/1.22.1/gstreamer-1.22.1.tar.gz"
+    sha256 "e8fc50d9f68c83bea3c7f04d78de2a9d54511c69b3449a7df8e4e2edf80f1e00"
+
+    # When updating this resource, use the tag that matches the GStreamer version.
+    resource "rs" do
+      url "https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/archive/gstreamer-1.22.1/gst-plugins-rs-gstreamer-1.22.1.tar.gz"
+      sha256 "65a8d76fda5f8b4a4ce088c574d055a823e9955352e1c6415e644dca4a655bae"
+
+      # Fix build failure on macOS.
+      # TODO: Upstream this (if it works).
```
Comment (carlocab) [inline Formula/gstreamer.rb:17]: Upstreamed at https://gitlab.freedesktop.org/gstreamer/gst-plugins-rs/-/issues/332. Restrictions on new accounts at freedesktop.org prevent me from sending an MR. Will update this PR with the link.

**Review by carlocab — COMMENTED** [review]
Ah, wait. This will need a whole lot of `link_overwrite`s, probably.

File: Formula/aravis.rb [hunk 2]
```diff
@@ -34,20 +32,26 @@ class Aravis < Formula
...
+        export GST_PLUGIN_PATH=#{opt_lib}/gstreamer-1.0
+    EOS
+  end
+
   test do
-    lib_ext = OS.mac? ? "dylib" : "so"
-    output = shell_output("gst-inspect-1.0 #{lib}/gstreamer-1.0/libgstaravis.#{version.major_minor}.#{lib_ext}")
+    output = shell_output(
+      "gst-inspect-1.0 #{lib}/gstreamer-1.0/#{shared_library("libgstaravis", version.major_minor)}",
+    )
```
Comment (carlocab) [inline Formula/aravis.rb:54]: I should revert this change -- the original was correct.

File: Formula/gstreamer.rb [hunk 3]
```diff
@@ -21,51 +37,221 @@ class Gstreamer < Formula
...
+  depends_on "gtk4"
+  depends_on "jpeg-turbo"
+  depends_on "json-glib"
+  depends_on "lame"
+  depends_on "libogg"
+  depends_on "libpng"
+  depends_on "libpthread-stubs"
+  depends_on "libshout"
+  depends_on "libsoup"
+  depends_on "libusrsctp"
```
Comment (carlocab) [inline Formula/gstreamer.rb:77]: This provides only a static library so there is no linkage. Will fix that separately.

**Review by chenrui333 — DISMISSED** [review]

**Review by p-linnane — APPROVED** [review]

File: Formula/gstreamer.rb [hunk 4]
```diff
@@ -1,71 +1,262 @@
...
-    # https://bugzilla.gnome.org/show_bug.cgi?id=750367
-    args = std_meson_args + %w[
+    (buildpath/"subprojects/gst-plugins-rs").install resource("rs")
+    site_packages = Language::Python.site_packages(python3)
+    # To pass arguments to subprojects (e.g. `gst-editing-services`), use
+    #   -Dsubproject:option=value
+    args = %W[
+      -Dpython=enabled
+      -Dlibav=enabled
+      -Dlibnice=disabled
```
Comment (nirbheek) [inline Formula/gstreamer.rb:133]: libnice is actually a really important gstreamer plugin, since it's essential for webrtc support.

I am not sure how you can solve the circular dep though. Maybe you could move libnice into the monorepo too? Is there anything else in Homebrew that uses libnice?

File: Formula/gstreamer.rb [hunk 5]
```diff
@@ -1,71 +1,262 @@
...
+    # The apple media plug-in uses API that was added in Mojave
+    args << "-Dgst-plugins-bad:applemedia=disabled" if MacOS.version <= :high_sierra
+
+    # Ban trying to chown to root.
+    # https://bugzilla.gnome.org/show_bug.cgi?id=750367
+    args << "-Dgstreamer:ptp-helper-permissions=none"
 
-    bin.env_script_all_files libexec/"bin", GST_PLUGIN_SYSTEM_PATH: HOMEBREW_PREFIX/"lib/gstreamer-1.0"
+    # Prevent the build from downloading an x86-64 version of bison.
+    inreplace "meson.build", "subproject('macos-bison-binary')", ""
```
Comment (nirbheek) [inline Formula/gstreamer.rb:179]: Ah, hmm, we should provide a way to disable this. I'll fix it upstream :)

**Review by nirbheek — COMMENTED** [review]
This is great stuff!

File: Formula/gstreamer.rb (hunk 5 above)
Comment (nirbheek) [reply Formula/gstreamer.rb:179]: https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/4428

You can start using this with the 1.24 release.

File: Formula/gstreamer.rb [hunk 6]
```diff
@@ -1,71 +1,262 @@
...
-      "cdata.set_quoted('PLUGINDIR', join_paths(get_option('prefix'), get_option('libdir'), 'gstreamer-1.0'))",
-      "cdata.set_quoted('PLUGINDIR', '#{HOMEBREW_PREFIX}/lib/gstreamer-1.0')"
-
-    mkdir "build" do
-      system "meson", *args, ".."
-      system "ninja", "-v"
-      system "ninja", "install", "-v"
-    end
+    # The apple media plug-in uses API that was added in Mojave
+    args << "-Dgst-plugins-bad:applemedia=disabled" if MacOS.version <= :high_sierra
```
[the remark this answers is not in the document]
Comment (carlocab) [reply Formula/gstreamer.rb:172]: This line originates from 03d561ad44f2, with associated PR at #52730. We don't test on systems older than Big Sur anymore, so I guess this has been fixed since then.

File: Formula/gstreamer.rb (hunk 5 above)
Comment (carlocab) [reply Formula/gstreamer.rb:179]: Thanks!

File: Formula/gstreamer.rb (hunk 4 above)
Comment (carlocab) [reply Formula/gstreamer.rb:133]: > Maybe you could move libnice into the monorepo too?

Yea, maybe. It'll make updating libnice really annoying, though. It's also not clear to me that you get the same libnice if you build it before Gstreamer rather than after, which might be an issue for users.

> Is there anything else in Homebrew that uses libnice?

Not currently, no.

Comment (nirbheek) [reply Formula/gstreamer.rb:133]: The monorepo will pull in libnice in the same way that it pulls in gst-plugins-rs (both are separate repos that are meson wraps), so it will be identical to building it separately. If nothing else is using libnice, then I suggest merging it into the monorepo too.

Note that libnice is maintained by Olivier, who is a gstreamer core maintainer, and I don't expect anyone outside of gstreamer to use the project.

If you really want to keep it separate, I'd recommend building it twice: once as its own project without gstreamer support, then again as part of the monorepo to build only the gstreamer plugin. That'll require non-trivial build system patching, but it can be done.

Comment (carlocab) [reply Formula/gstreamer.rb:133]: It's not quite the same as gst-plugins-rs, though, which kind of supports updating simultaneously with gstreamer.

If we build libnice with gstreamer, then we need to rebuild all of gstreamer every time we want to update only libnice (if we even remember to update it after it becomes part of gstreamer).

Comment (nirbheek) [reply Formula/gstreamer.rb:133]: The libnice wrap file is continuously updated, so you'd get updates whenever you update gstreamer (stable releases for instance). We test the two as a unit in the monorepo and in cerbero.

Comment (carlocab) [reply Formula/gstreamer.rb:133]: Only getting updates when gstreamer is updated is also not really desirable.

Comment (nirbheek) [reply Formula/gstreamer.rb:133]: fwiw, gst-plugins-rs is on a completely different release cycle than the rest of gstreamer, so you're already doing that for the rust plugins.

Comment (carlocab) [reply Formula/gstreamer.rb:133]: Yes, but hacking up gst-*-rs so that it can work when updated independently isn't really something we want to do (though avoiding that introduced the need for other workarounds...). libnice seems to still be usable when updated independently.

Comment (nirbheek) [reply Formula/gstreamer.rb:133]: I would argue that if nothing else in Homebrew is using libnice, then it's not really usable in its current state since gstreamer can only use libnice if the gstreamer plugin is also available :slightly_smiling_face: 

Like I said, we can fix the circular dep by building only the gstreamer plugin as part of the monorepo build, do you want me to help you with that?

Comment (carlocab) [reply Formula/gstreamer.rb:133]: :shrug: It has at least ~40k downloads in the past year, with zero complaints about it being unusable, so I'd hazard that some users are using it for _something_ even if it doesn't really work with gstreamer.

Let's leave it this way for the moment. If multiple users complain about not being able to use libnice, then we can look into including it in the monorepo build.

Comment (nirbheek) [reply Formula/gstreamer.rb:133]: Actually, we get all the complaints about that :) people complain that the webrtc plugin doesn't work in Homebrew, and we have to tell them to use the official binaries instead.

Gstreamer's webrtc support is probably the most widely-used feature out there right now (other than plain playback).

Comment (carlocab) [reply Formula/gstreamer.rb:133]: Send them here the next time they come around then, and we'll take a look.

**Comment by spiegela** [comment]
Interesting, when installing this version, `gstreamer` pulls in both `gtk+3` and `gtk+4` for me, but then `gst-inspect-1.0` complains about them both being installed:

```
objc[52356]: Class ResultReceiver is implemented in both /opt/homebrew/Cellar/gtk+3/3.24.37/lib/libgtk-3.0.dylib (0x106cd5df8) and /opt/homebrew/Cellar/gtk4/4.10.1/lib/libgtk-4.1.dylib (0x1159c2ec0). One of the two will be used. Which one is undefined.
``

File: Formula/gstreamer.rb (hunk 4 above)
Comment (cooltea713705) [reply Formula/gstreamer.rb:133]: Hi! Thanks for documenting the issue: upgrading to 1.22.2 effectively made `webrtcbin` no longer be listed as plugin. I understand the circular dependency issue, can some work be done to include `libnice` so that `webrtcbin` can be supported? It's not clear what the best approach is if `libnice` still needs to be a standalone.

Comment (carlocab) [reply Formula/gstreamer.rb:133]: Did webrtc work for you before? @nirbheek tells me that it was basically broken previously.

Comment (cooltea713705) [reply Formula/gstreamer.rb:133]: Up until 1.20.0 at least what it took was for gstreamer + gst-plugins-bad to be installed via brew in which case it was working yes.

Comment (nirbheek) [reply Formula/gstreamer.rb:133]: Earlier, all plugin registration was broken unless you set the right env vars, and linking was also broken. webrtcbin was not _specifically_ broken.

I can help you with a solution for the circular dep, just let me know if you need my help.

One way would be to allow gstnice (inside libnice) to be built against a system libnice. The plugin only uses the version string at build time, nothing else.

I can patch that and talk to Olivier about upstreaming it.

Comment (carlocab) [reply Formula/gstreamer.rb:133]: > One way would be to allow gstnice (inside libnice) to be built against a system libnice. The plugin only uses the version string at build time, nothing else.

That seems like it could work.
